docs/auth.md

```diff
@@ -0,0 +1,147 @@
+## User Authentication in Soul
+
+Soul incorporates a robust user authentication system that handles user accounts, groups, permissions, and cookie-based user sessions. This section provides an overview of how the default implementation works.
+
+### Overview
+
+The Soul authentication system handles both authentication and authorization. Briefly, authentication verifies a user is who they claim to be, and authorization determines what an authenticated user is allowed to do. Here, the term "authentication" is used to refer to both tasks.
+
+The auth system consists of:
+
+- Users
+- Roles and Permissions: A generic way of applying labels and permissions to more than one user.
+- A password hashing system
+- APIs for logging in users or restricting content
+- JWT based access tokens
+
+The authentication system in Soul aims to be very generic and doesn't provide some features commonly found in web authentication systems, such as:
+
+- Password strength checking
+- Throttling of login attempts
+- Authentication against third-parties (OAuth, for example)
+- Object-level permissions
+
+![auth-db-diagram](./auth-db-diagram.png)
+
+<center style="font-size: 12px; color: gray">Figure 1: Soul Authentication Database Diagram</center>
+
+#### Users
+
+The `_users` table serves as the central component of the authentication system. Each row in the `_users` table represents an individual who interacts with your Soul instance. This table is crucial for enabling various functionalities, such as access restriction, and user registration.
+
+**_Superusers_**, distinguished by the `is_superuser` attribute set to true, are a special type of user in Soul. They can bypass any permissions and restrictions, effectively granting them "God-mode" within the system. Authorization limitations do not apply to superusers.
+
+The attributes of the `_users` table are:
+
+- id int
+- username varchar
+- \_hashed_password varchar
+- \_salt varchar
+- is_superuser boolean
+- created_at datetime
+- updated_at datetime
+
+Note that when Soul boots up, it looks for a table called "\_users" which holds the Users mentioned above.
+
+##### Modifying Superusers
+
+Due to the sensitive nature of the superuser status, it is not possible to change the `is_superuser` attribute of a user through the API.
+
+Instead you can create, update and delete superusers using the command line.
```

Comment:
@thevahidal - it looks like these commands just take `username` and `password` fields.  

1. Shouldn't the command take the `id` of the record being edited?
2. How do we edit the `id_superuser` field?  I _think_ it should be part of this custom route.

Reply:
1. As the username is unique, I think it's more readable and therefore less prone to mistakenly update/remove unwanted users.
2. Right now we only can edit its password. What other fields you're thinking about updating?

Reply:
The id is the primary key, this means we can change the username (such as when a staff member changes surname on marriage). I don't think this command-line API allows for a new username to be provided.

Reply:
Sure that's a good point, maybe we can change it to take `id` instead.

Reply:
Yes, let's.  Well, it _may_ still take username, but that's part of updating the row, rather than identifying which row to update.

Reply:
Hmm, I was about to step in and implement these documentation changes, but there may be a simpler strategy.

It's just the `is_superuser` attribute that we need to protect.  So, how about we let the normal API handle superuser updates, with the exception of the `is_superuser` flag, or the `password` for super-users :
`updatesuperuser --userid=12 --is_superuser=false`
`updatesuperuser --userid=12 --is_superuser=true --password=pwd`
`updatesuperuser --userid=12 --password=pwd`

I _think_ this still gives us the protection we require around superusers, and reduces the number of commands we have to add.

Reply:
https://github.com/thevahidal/soul/blob/auth-docs/docs/auth.md#modifying-superusers

Reply:
Thanks for the clarification @thevahidal .
But, I don’t see superusers as a custom type of user - it’s just the setting of that flag.  If we use the special command to clear the `is_superuser` flag then they are no longer a superuser, and we can delete it via the REST API.  Similarly, we could create a normal user using REST, then use the special command to set `is_superuser` to true.
That suggests we only need to _protect_ the `is_superuser` flag, and the passwords for anyone with `is_superuser`.

Reply:
Since you added the option to promote and demote a user to / from superuser I thought it's better to call this command "[action]user" instead of "[action]supueruser".

It's basically the same thing and though you're right that we can modify normal users using REST, it's just another way of doing it with a higher level of access.

Reply:
Still I don't have an strong opinion on this one, It's just a matter of naming and we can revert it to the previous one.

Reply:
So you mean we don't need the createuser and deleteuser, is that right?

Reply:
Yes, that's right.

All operations go through REST, except for the specific changes related to is_superuser.